store/postgres/src/metadata.rs

```diff
@@ -0,0 +1,166 @@
+        block_range -> Range<Integer>,
+    }
+}
+
+table! {
+    subgraphs.subgraph_manifest (vid) {
+        vid -> BigInt,
+        id -> Text,
+        spec_version -> Text,
+        description -> Nullable<Text>,
+        repository -> Nullable<Text>,
+        schema -> Text,
+        data_sources -> Array<Text>,
+        templates -> Nullable<Array<Text>>,
+        block_range -> Range<Integer>,
+    }
+}
+
+table! {
+    subgraphs.ethereum_contract_data_source (vid) {
+        vid -> BigInt,
+        id -> Text,
+        kind -> Text,
+        name -> Text,
+        network -> Nullable<Text>,
+        source -> Text,
+        mapping -> Text,
+        templates -> Nullable<Array<Text>>,
+        block_range -> Range<Integer>,
+    }
+}
+
+/// Look up the graft point for the given subgraph in the database and
+/// return it
+pub fn deployment_graft(
+    conn: &PgConnection,
+    id: &SubgraphDeploymentId,
+) -> Result<Option<(SubgraphDeploymentId, EthereumBlockPointer)>, StoreError> {
+    use subgraph_deployment as sd;
+
+    if id.is_meta() {
+        // There is no SubgraphDeployment for the metadata subgraph
+        Ok(None)
+    } else {
+        match sd::table
+            .select((sd::graft_base, sd::graft_block_hash, sd::graft_block_number))
+            .filter(sd::id.eq(id.as_str()))
+            .first::<(Option<String>, Option<Vec<u8>>, Option<BigDecimal>)>(conn)?
+        {
+            (None, None, None) => Ok(None),
+            (Some(subgraph), Some(hash), Some(block)) => {
+                let hash = H256::from_slice(hash.as_slice());
+                let block = block.to_u64().expect("block numbers fit into a u64");
+                let subgraph = SubgraphDeploymentId::new(subgraph.clone()).map_err(|_| {
+                    StoreError::Unknown(format_err!(
+                        "the base subgraph for a graft must be a valid subgraph id but is `{}`",
+                        subgraph
+                    ))
+                })?;
+                Ok(Some((subgraph, EthereumBlockPointer::from((hash, block)))))
+            }
+            _ => unreachable!(
+                "graftBlockHash and graftBlockNumber are either both set or neither is set"
+            ),
+        }
+    }
+}
+
+pub fn subgraph_schema(
+    conn: &PgConnection,
+    id: SubgraphDeploymentId,
+) -> Result<Schema, StoreError> {
+    // The subgraph of subgraphs schema is built-in and doesn't have a
+    // SubgraphManifest in the database
+    const SUBGRAPHS_SCHEMA: &str = include_str!("subgraphs.graphql");
+    let res = if id.is_meta() {
+        Schema::parse(SUBGRAPHS_SCHEMA, id)
+    } else {
+        use subgraph_manifest as sm;
+        let manifest_id = SubgraphManifestEntity::id(&id);
+        let s: String = sm::table
+            .select(sm::schema)
+            .filter(sm::id.eq(manifest_id.as_str()))
+            .first(conn)?;
+        Schema::parse(s.as_str(), id)
+    };
+    res.map_err(|e| StoreError::Unknown(e))
+}
```

Comment:
It's so you can get at these things without needing a `Store`. The way it was before introduced a circular dependency where a `Store` needed an `entities::Connection`, and to construct an `entities::Connection` you needed a `Store`. The latter was solely because the `Store` is also a cache of schemas. 

That leads to two complications: (1) we now want the `Store` to also cache information about a possible graft point since that needs to be checked on every revert; when we create a grafted subgraph there's a point where we need an `entities::Connection` before the graft information has been written to the database. If we go through the cache, we could end up storing incorrect information in the cache, or need to remember to invalidate the cache entry in certain cases. It's more robust to have a way to get the schema that bypasses the cache when we are not sure things are truly cacheable yet. (2) It causes trouble with @That3Percent's `with_entity_connection` API for similar reasons (i.e., the seeming circularity of the `Store` and `entities::Connection` dependencies because of caching)

Reply:
Ok, got it!